test/test_autograd.py

```diff
@@ -9524,6 +9532,128 @@ def fn(x1, x2):
+                                              func=schema)
+
+        derivative = load_derivatives.create_derivative(
+            native_function,
+            formula='func_backward(grad_x, grad_y)',
+            var_names=(),
+            available_named_gradients=['grad_x', 'grad_y'])
+        self.assertSetEqual(derivative.named_gradients, {'grad_x', 'grad_y'})
+
+    def test_non_differentiable_output(self):
+        specification = 'func(Tensor a, Tensor b) -> (Tensor x, bool y, Tensor z)'
+        schema = tools.codegen.model.FunctionSchema.parse(specification)
+        native_function = dataclasses.replace(DEFAULT_NATIVE_FUNCTION,
+                                              func=schema)
+
+        differentiability_info = load_derivatives.create_differentiability_info(
+            defn={'name': specification,
+                  'a': 'grads[0]',
+                  'b': 'grads[2]',
+                  },
+            functions_by_signature={schema.signature(): [native_function]},
+            functions_by_schema={specification: native_function},
+        )
+
+        self.assertListEqual(differentiability_info.available_named_gradients,
+                             # grad_y is not present because y is a
+                             # bool and thus not differentiable.
+                             ['grad_x', 'grad_z'])
+
+    def test_indexed_grads(self):
+        schema = tools.codegen.model.FunctionSchema.parse(
+            'func(Tensor a, Tensor b) -> (Tensor x, Tensor y)')
+        native_function = dataclasses.replace(DEFAULT_NATIVE_FUNCTION,
+                                              func=schema)
+
+        derivative = load_derivatives.create_derivative(
+            native_function,
+            formula='func_backward(grads[0], grads[1])',
+            var_names=(),
+            available_named_gradients=['grad_x', 'grad_y'])
+        self.assertSetEqual(derivative.named_gradients, set())
+
+    def test_named_grads_and_indexed_grads(self):
+        specification = 'func(Tensor a, Tensor b) -> (Tensor x, Tensor y)'
+        schema = tools.codegen.model.FunctionSchema.parse(specification)
+        native_function = dataclasses.replace(DEFAULT_NATIVE_FUNCTION,
+                                              func=schema)
+
+        with self.assertRaisesRegex(RuntimeError,
+                                    'illegally mixes use of "grad_RETURN_NAME"'):
+            load_derivatives.create_differentiability_info(
+                defn={'name': specification,
+                      # Uh-oh, the derivatives reference gradients by
+                      # name and by index.
+                      'a': 'grad_x',
+                      'b': 'grads[1]',
+                      },
+                functions_by_signature={schema.signature(): [native_function]},
+                functions_by_schema={specification: native_function},
+            )
+
+class TestGenAutogradFunctions(TestCase):
```

Comment:
Why do you use two classes here? One for all the codegen stuff should be enough?

Reply:
Note: moved to its own file now.

I was thinking one test case per module. No good?